components/libraryprovider.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python3
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import yaml
+
+from components.utilities import Struct
+from components.providerbase import BaseProvider, INeedsCommandProvider, INeedsLoggingProvider
+
+
+# Library metadata from moz.yaml files that we care about and example values
+# NOTE: yaml_path is provided at import and may change based on the gecko-path
+#       given in the localconfig
+# NOTE: this format mirrors the moz.yaml schema defined at
+#       https://searchfox.org/mozilla-central/source/python/mozbuild/mozbuild/vendor/moz_yaml.py
+
+# 'bugzilla': {
+#     'product': 'Core',
+#     'component': 'ImageLib'
+# },
+# 'origin': {
+#     'name': 'dav1d'
+# },
+# 'updatebot': {
+#     'enabled': True,
+#     'maintainer-bz': 'nobody@mozilla.com',
+#     'maintainer-phab': 'nobody'
+# },
+# 'yaml_path': '/Users/nobody/mozilla-central/media/libdav1d/moz.yaml'
+
+
+class LibraryProvider(BaseProvider, INeedsCommandProvider, INeedsLoggingProvider):
+    def __init__(self, config):
+        pass
+
+    def get_libraries(self, gecko_path):
+        libraries = []
+        mozilla_central_yamls = self.run(["find", gecko_path, "-type", "f", "-name", "moz.yaml"]).stdout.decode().strip().split("\n")
+
+        for file in mozilla_central_yamls:
+            with open(file, "r") as mozyaml:
+                new_library = yaml.safe_load(mozyaml.read())
+                new_library['yaml_path'] = file
+                libraries.append(self.validate_library(new_library))
+        return libraries
+
+    def validate_library(self, library):
```

Comment:
I think this could be refactored better.

1: When you create an object where you get data out like library.origin["name"] - you're translating the moz.yaml structure into the object. Why does the updatebot code care about the moz.yaml format? It seems needlessly verbose/complex.  Why not just library.name ?
2: The pattern below is heavily repeated.  `if foo in library and bar in library["foo"]: assign else: raise`  It seems like a very strong candidate for turning those four lines into a sub-function and then calling it in different situations.

That said, we don't need to do this refactoring now, if at all.